proposals/crd-support.md

```diff
@@ -0,0 +1,52 @@
+
+## Motivation
+The current ways to handle CRDs in Helm have issues and these are known to have caused real problems for users (more details in [#4863](https://github.com/helm/helm/issues/4863)). The goal here is to handle CRDs in a way that is intuitive and flexible for chart developers, as well as easy to use and safe for helm users. The goal is that this could be introduced in Helm 2 as well as make it into Helm 3.
+
+## Goals
+* The solution should not be dependent on versioned CRDs. Even though these are in development, a solution needs to work with existing CRDs. But we would like a solution that can work with/take advantage of versioned CRDs when they become available.
+* It should support existing installed charts that use CRDs, regardless of which of the existing ways to install CRDs was used.
+* CRDs should never be deleted unless the user has explicitely asked for it. Since CRDs are global in a Kubernetes cluster, we can never know for sure that it is no longer in use.
+* CRDs should never be updated unless the user has explicitely asked for it. Updating CRDs could render them unusable for some of the services/applications that depend on them.
+* Mixing CRDs and CRs of the type in the same chart should be supported.
+
+
+## Implementation
+This proposal takes a very conservative approach on any changes to CRDs, but the hope is that it will not be to inconvenient. If CRDs are either only used by a single chart, or is used by many charts but rarely updated (like App CRD), updates should usually be possible without too much inconvenience for the user.
+
+CRDs need to be handled correctly through all the phases of a Helm release:
+
+### Install
+* CRDs should be the first resource processed during installation. CRDs must be installed into the cluster before any CRs belonging to the CRD.
+* If a chart contains one or more CRDs, the helm install process should for each of them, check if the CRD is already installed. 
+  * If it is, Helm should check if the installed CRD is identical to the one contained in the chart. 
+    * If it is, the install process can skip installing the CRD and just continue.
+    * If it is different, the install process should fail, unless the user has explicitly confirmed through a command line flag that the CRD in the chart should overwrite the existing one. If a CRD is versioned, it should allow multiple different versions of the CRD to coexist in the cluster. Overwrite will only apply if a chart has a CRD with a version that already exists.
+  * If it does not already exist, the CRD will be installed.
+* All CRDs in the chart will be tracked as part of a release when the chart is installed
+  * If the CRD was already installed, then the existing CRD will be adopted into the release.
```

Comment:
What if two charts install a CRD? The instance of the first would track it in an release. The second chart would install and adopt it into the release. It would then be part of multiple releases?

Should we instead treat it as a global object (which is how k8s sees it)? What benefit do we get from tracking it?

Reply:
I don't understand the implications of adopting a CRD. It seems like this could get very confusing if a chart installs multiple CRDs. Imagine this case:

Chart A installs CRD X and CRD Y
Chart B Installs CRD X and CRD Z

According to your proposal, now both A and B "own" CRD X. And there is no way to delete CRD Y or CRD Z except by forcing a deletion of all CRDs. 

So assume you do a `helm delete --delete-crds` on Chart B. Now you've deleted CRD X out from under Chart A. But if you don't do it that way, there is no way to delete CRD Y.

But requiring users to explicitly indicate which CRDs they want to uninstall requires users to get REALLY familiar with their charts (which, IMO, is unfair).

Now, if you skip the adoption part, then only by uninstalling Chart A can you delete CRD X. That is at least somewhat clear, though it only sidesteps part of the problem, as deleting A with `--delete-crds` would render B useless. (And yet, still, no way to delete CRD Y from A).

Adopting ANY object is very dangerous, and I suggest avoiding it as an automated behavior simply because it is counterintuitive. (Consider if your OS package manager decided to adopt parts of a program that were installed elsewhere)

Reply:
Yeah, I see your point here. The thought behind adopting an already existing CRD into a release was to have a way to keep track of which resources (including CRDs) that a release has created/depends on. CRDs are different than other resources, so it does get a little bit confusing.
I think the alternative to trying to keep track of which releases uses each CRD, is the path taken with the crd_install hook where CRDs are installed and "forgotten" by the release. It would have the drawback that CRDs will be left in your cluster after any charts that used it has been deleted, but that might be the most likely outcome even with this proposal.
Tracking CRDs only as part of the release that initially created it seems impractical and confusing.